### DISCUSSION TITLE
TypeScript: Type checking RHF methods in child components

### DISCUSSION THREAD

**Discussion by mrlubos** [react-hook-form/react-hook-form#2625] [Q&A] [answered]
Hey! I'm defining a new form in a React component.

```ts
const { errors } = useForm<FormShape>();
```

And I want to pass the `errors` prop to the child component.

```tsx
function ParentComponent() {
  const { errors } = useForm<FormShape>();

  return (
    <ChildComponent errors={errors} />
  )
}
```

I'd like to have the same type checking features I get in `ParentComponent` available in the `ChildComponent`. I cannot write my own type because I'd need to know the `FormShape` type in advance. However, `ChildComponent` can be rendered by multiple parent components, so it's not possible to know all possible shapes.

Is it possible to achieve the same type-checking features in the `ChildComponent` as in the `ParentComponent`? Every solution I tried writing my own type failed so far. Thanks!

**Comment by bluebill1049** [comment, marked as answer]
Do you have a codesandbox to demonstrate your question?

**Reply by mrlubos** [reply]
Hi @bluebill1049! Sure, here's an [example](https://codesandbox.io/s/react-hook-form-typescript-forked-v36c7). Notice TypeScript complains in the `Child` component that `foo` does not exist on the object. That makes sense as it's not explicitly defined in props. My question is, how to get the same type-checking features as I have in `Parent` where autocomplete works as expected? Thank you!

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/react-hook-form-typescript-forked-7671s

**Reply by mrlubos** [reply]
Yes, that works for a single `Parent` component, but it needs to work with multiple form shapes which I don't know in advance, see `Parent2` in this [example](https://codesandbox.io/s/react-hook-form-typescript-forked-x0fwn).

To be fair, I only really need access to the error object and type-checking on that. I managed to get that working by using `any` on the `errors` object and type casting the result of `errors[key]` to `FieldError` from `react-hook-form`. Still, it doesn't feel right to disable type-checking for that prop

**Reply by bluebill1049** [reply]
what do you mean? you can export your FormValues and reuse it, you type is broken because you set errors to objects.

**Reply by cpprookie** [reply]
I think @mrlubos  wants to build a `Child` component which is reusable for different  `Parent` Forms @bluebill1049. What you actually needs is [generic types for tsx](https://mariusschulz.com/articles/passing-generics-to-jsx-elements-in-typescript).  But it has nothing to  do with the react-hook-form library. Maybe render-props is suitable in this situation ? You can  Get `error`  in `Parent` context then do anything you want.

**Reply by mrlubos** [reply]
Hi @cpprookie, that's correct. I'll check the article you linked, thank you!

**Reply by cpprookie** [reply]
You're welcome @mrlubos . Hope it could help you.
